src/ray/object_manager/object_manager.cc

```diff
@@ -655,4 +666,47 @@ void ObjectManager::ExecuteReceiveObject(const ClientID &client_id,
                  << "/" << config_.max_receives;
 }
 
+void ObjectManager::ReceiveFreeRequest(std::shared_ptr<TcpClientConnection> &conn,
+                                       const uint8_t *message) {
+  auto fr = flatbuffers::GetRoot<object_manager_protocol::FreeRequestMessage>(message);
+  std::vector<ObjectID> object_ids = from_flatbuf(*fr->object_ids());
+  // This RPC should come from another Object Manager.
+  // Do not spread this request.
+  bool spread = false;
+  FreeObjects(object_ids, spread);
+}
+
+void ObjectManager::FreeObjects(const std::vector<ObjectID> &object_ids, bool spread) {
+  buffer_pool_.FreeObjects(object_ids);
+  if (spread) {
+    SpreadFreeObjectRequest(object_ids);
+  }
+}
+
+void ObjectManager::SpreadFreeObjectRequest(const std::vector<ObjectID> &object_ids) {
+  // This code path should be called from node manager.
+  flatbuffers::FlatBufferBuilder fbb;
+  flatbuffers::Offset<object_manager_protocol::FreeRequestMessage> request =
+      object_manager_protocol::CreateFreeRequestMessage(fbb, to_flatbuf(fbb, object_ids));
+  fbb.Finish(request);
+  auto callback = [this, &fbb](const RemoteConnectionInfo &connection_info) {
+    ray::Status status;
+    std::shared_ptr<SenderConnection> conn;
+    status = connection_pool_.GetSender(ConnectionPool::ConnectionType::MESSAGE,
+                                        connection_info.client_id, &conn);
+    RAY_CHECK_OK(status);
+    if (conn == nullptr) {
+      conn = CreateSenderConnection(ConnectionPool::ConnectionType::MESSAGE,
+                                    connection_info);
+      connection_pool_.RegisterSender(ConnectionPool::ConnectionType::MESSAGE,
+                                      connection_info.client_id, conn);
+    }
+    RAY_CHECK_OK(conn->WriteMessage(
```

Comment:
Currently, `RAY_CHECK_OK` will directly fail if the condition is not satisfied. If you are working to change this to throwing exception, one overhead is that the connection cannot be released by `connection_pool_.ReleaseSender`. If that is the case, we can use the mechanism like shared pointer.

Reply:
We strictly do not use C++ exceptions. We can propagate status codes, but in this case, I think if we fail to write a message to a remote node manager we can just ignore it and not send the message.

@elibol should we clean up the connection in that case?

Reply:
@robertnishihara I think we can just ignore the failure assuming the proper functioning of Ray applications do not depend on this write operation. I think this is true since we're just removing objects from the object store.

Note that the `ReleaseSender` method just releases the connection from the connection pool. We'll need a method similar to `RemoveReceiver`, which, in addition to removing the connection from the sender connection maps, will also need to remove connections from the `available_*` sender maps. Once connections are no longer referenced, I expect them to be closed (we ought to confirm via asio documentation), but it might make sense to invoke close explicitly so that any subsequent attempt to use the connection fails. If we close the connection explicitly, we might want to hint at such semantics by naming the method something like `DestroySender` or `RemoveAndCloseSender`.

We could simplify the connection pool API by converting the `RemoveReceiver` method to a `RemoveConnection` method; the lookups required to figure out connection type etc. are cheap.

Reply:
@elibol Thanks for the confirmation. I will remove `RAY_CHECK_OK` for `WriteMessage`.

Reply:
Since `ReleaseSender ` is only declared but not implemented. I add a `TODO` comment and leave this to another PR.